[Bug]: Crash in MediaDetailPagerFragment due to lack of no-arg constructor

### Summary

We've got a whole bunch of crash reports with a cause that seems to surround with instantiating the MediaDetailPagerFragment. One some exploration the cause seems to be due to the lack of a public default constructor in MediaDetailPagerFragment. There's a private no-arg constructor but Android needs a public no-arg contructor to instantiate fragments [[ref](https://stackoverflow.com/a/16977633/5614968)].

### Steps to reproduce

I don't have specific steps. @nicolas-raoul could you share a set of steps if possible ?

### Expected behaviour

The MediaDetailPagerFragment has  a public no-arg constructor for Android to use.

### Actual behaviour

Instantiating the MediaDetailPagerFragment fails when Android is trying to restore it. We need to introduce a public no-arg constructor and possibly avoid the parameterized constructors for the fragment by receive the corresponding parameters via Bundle arguments as suggested in [this Q&A](https://stackoverflow.com/a/16977633/5614968)

### Device name

_No response_

### Android version

13

### Commons app version

4.1.0

**Note**

This does not seem specific to 4.1.0 or Android 13. There are crash reports from other versions like 4.0.3 and Android versions like 10.

### Device logs

 The trace of one such crash looks as follows:

<details><summary>Stack trace of MediaDetailPagerFragment crash</summary>
<p>

```stacktrace
java.lang.RuntimeException: Unable to start activity ComponentInfo{fr.free.nrw.commons/fr.free.nrw.commons.contributions.MainActivity}: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment fr.free.nrw.commons.media.MediaDetailPagerFragment: could not find Fragment constructor
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3644)
at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3781)
at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5782)
at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5673)
at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:71)
at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:138)
at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:95)
at android.app.ClientTransactionHandler.executeTransaction(ClientTransactionHandler.java:61)
at android.app.ActivityThread.handleRelaunchActivityLocally(ActivityThread.java:5741)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2316)
at android.os.Handler.dispatchMessage(Handler.java:106)
at android.os.Looper.loopOnce(Looper.java:201)
at android.os.Looper.loop(Looper.java:288)
at android.app.ActivityThread.main(ActivityThread.java:7918)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:936)
Caused by: androidx.fragment.app.Fragment$InstantiationException: Unable to instantiate fragment fr.free.nrw.commons.media.MediaDetailPagerFragment: could not find Fragment constructor
at androidx.fragment.app.Fragment.instantiate(Fragment.java:628)
at androidx.fragment.app.FragmentContainer.instantiate(FragmentContainer.java:57)
at androidx.fragment.app.FragmentManager$3.instantiate(FragmentManager.java:483)
at androidx.fragment.app.FragmentStateManager.<init>(FragmentStateManager.java:85)
at androidx.fragment.app.FragmentManager.restoreSaveState(FragmentManager.java:2728)
at androidx.fragment.app.Fragment.restoreChildFragmentState(Fragment.java:1890)
at androidx.fragment.app.FragmentStateManager.create(FragmentStateManager.java:479)
at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:278)
at androidx.fragment.app.FragmentStore.moveToExpectedState(FragmentStore.java:112)
at androidx.fragment.app.FragmentManager.moveToState(FragmentManager.java:1647)
at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3128)
at androidx.fragment.app.FragmentManager.dispatchCreate(FragmentManager.java:3061)
at androidx.fragment.app.FragmentController.dispatchCreate(FragmentController.java:240)
at androidx.fragment.app.FragmentActivity.onCreate(FragmentActivity.java:276)
at androidx.appcompat.app.AppCompatActivity.onCreate(AppCompatActivity.java:106)
at fr.free.nrw.commons.di.CommonsDaggerAppCompatActivity.onCreate(CommonsDaggerAppCompatActivity.java:24)
at fr.free.nrw.commons.theme.BaseActivity.onCreate(BaseActivity.java:28)
at fr.free.nrw.commons.contributions.MainActivity.onCreate(MainActivity.java:119)
at android.app.Activity.performCreate(Activity.java:8342)
at android.app.Activity.performCreate(Activity.java:8321)
at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1417)
at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3625)
... 17 more
Caused by: java.lang.NoSuchMethodException: fr.free.nrw.commons.media.MediaDetailPagerFragment.<init> []
at java.lang.Class.getConstructor0(Class.java:2363)
at java.lang.Class.getConstructor(Class.java:1759)
at androidx.fragment.app.Fragment.instantiate(Fragment.java:613)
... 38 more
```


</p>
</details> 

### Screen-shots

_No response_

### Would you like to work on the issue?

None

--- a/app/src/main/java/fr/free/nrw/commons/bookmarks/BookmarkListRootFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/bookmarks/BookmarkListRootFragment.java
@@ -184,7 +184,7 @@ public class BookmarkListRootFragment extends CommonsDaggerSupportFragment imple
     public void refreshNominatedMedia(int index) {
         if (mediaDetails != null && !listFragment.isVisible()) {
             removeFragment(mediaDetails);
-            mediaDetails = new MediaDetailPagerFragment(false, true);
+            mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
             ((BookmarkFragment) getParentFragment()).setScroll(false);
             setFragment(mediaDetails, listFragment);
             mediaDetails.showImage(index);
@@ -243,7 +243,7 @@ public class BookmarkListRootFragment extends CommonsDaggerSupportFragment imple
         Log.d("deneme8", "on media clicked");
         container.setVisibility(View.VISIBLE);
         ((BookmarkFragment) getParentFragment()).tabLayout.setVisibility(View.GONE);
-        mediaDetails = new MediaDetailPagerFragment(false, true);
+        mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
         ((BookmarkFragment) getParentFragment()).setScroll(false);
         setFragment(mediaDetails, listFragment);
         mediaDetails.showImage(position);

--- a/app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/category/CategoryDetailsActivity.java
@@ -115,7 +115,7 @@ public class CategoryDetailsActivity extends BaseActivity
         mediaContainer.setVisibility(View.VISIBLE);
         if (mediaDetails == null || !mediaDetails.isVisible()) {
             // set isFeaturedImage true for featured images, to include author field on media detail
-            mediaDetails = new MediaDetailPagerFragment(false, true);
+            mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
             FragmentManager supportFragmentManager = getSupportFragmentManager();
             supportFragmentManager
                     .beginTransaction()

--- a/app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/contributions/ContributionsFragment.java
@@ -668,7 +668,7 @@ public class ContributionsFragment
     @Override
     public void showDetail(int position, boolean isWikipediaButtonDisplayed) {
         if (mediaDetailPagerFragment == null || !mediaDetailPagerFragment.isVisible()) {
-            mediaDetailPagerFragment = new MediaDetailPagerFragment(false, true);
+            mediaDetailPagerFragment = MediaDetailPagerFragment.newInstance(false, true);
             if(isUserProfile) {
                 ((ProfileActivity)getActivity()).setScroll(false);
             }
@@ -749,7 +749,7 @@ public class ContributionsFragment
     public void refreshNominatedMedia(int index) {
         if(mediaDetailPagerFragment != null && !contributionsListFragment.isVisible()) {
             removeFragment(mediaDetailPagerFragment);
-            mediaDetailPagerFragment = new MediaDetailPagerFragment(false, true);
+            mediaDetailPagerFragment = MediaDetailPagerFragment.newInstance(false, true);
             mediaDetailPagerFragment.showImage(index);
             showMediaDetailPagerFragment();
         }

--- a/app/src/main/java/fr/free/nrw/commons/explore/ExploreListRootFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/ExploreListRootFragment.java
@@ -111,7 +111,7 @@ public class ExploreListRootFragment extends CommonsDaggerSupportFragment implem
     public void onMediaClicked(int position) {
         container.setVisibility(View.VISIBLE);
         ((ExploreFragment) getParentFragment()).tabLayout.setVisibility(View.GONE);
-        mediaDetails = new MediaDetailPagerFragment(false, true);
+        mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
         ((ExploreFragment) getParentFragment()).setScroll(false);
         setFragment(mediaDetails, listFragment);
         mediaDetails.showImage(position);

--- a/app/src/main/java/fr/free/nrw/commons/explore/ExploreMapRootFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/ExploreMapRootFragment.java
@@ -118,7 +118,7 @@ public class ExploreMapRootFragment extends CommonsDaggerSupportFragment impleme
     public void onMediaClicked(int position) {
         container.setVisibility(View.VISIBLE);
         ((ExploreFragment) getParentFragment()).tabLayout.setVisibility(View.GONE);
-        mediaDetails = new MediaDetailPagerFragment(false, true);
+        mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
         ((ExploreFragment) getParentFragment()).setScroll(false);
         setFragment(mediaDetails, mapFragment);
         mediaDetails.showImage(position);

--- a/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/SearchActivity.java
@@ -221,7 +221,7 @@ public class SearchActivity extends BaseActivity
         searchView.setVisibility(View.GONE);// to remove searchview when mediaDetails fragment open
         if (mediaDetails == null || !mediaDetails.isVisible()) {
             // set isFeaturedImage true for featured images, to include author field on media detail
-            mediaDetails = new MediaDetailPagerFragment(false, true);
+            mediaDetails = MediaDetailPagerFragment.newInstance(false, true);
             supportFragmentManager
                     .beginTransaction()
                     .hide(supportFragmentManager.getFragments().get(supportFragmentManager.getBackStackEntryCount()))

--- a/app/src/main/java/fr/free/nrw/commons/explore/depictions/WikidataItemDetailsActivity.java
+++ b/app/src/main/java/fr/free/nrw/commons/explore/depictions/WikidataItemDetailsActivity.java
@@ -153,7 +153,7 @@ public class WikidataItemDetailsActivity extends BaseActivity implements MediaDe
         mediaContainer.setVisibility(View.VISIBLE);
         if (mediaDetailPagerFragment == null || !mediaDetailPagerFragment.isVisible()) {
             // set isFeaturedImage true for featured images, to include author field on media detail
-            mediaDetailPagerFragment = new MediaDetailPagerFragment(false, true);
+            mediaDetailPagerFragment = MediaDetailPagerFragment.newInstance(false, true);
             FragmentManager supportFragmentManager = getSupportFragmentManager();
             supportFragmentManager
                     .beginTransaction()

--- a/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
@@ -60,7 +60,7 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
     private static CompositeDisposable compositeDisposable = new CompositeDisposable();
 
     @BindView(R.id.mediaDetailsPager) ViewPager pager;
-    private Boolean editable;
+    private boolean editable;
     private boolean isFeaturedImage;
     private boolean isWikipediaButtonDisplayed;
     MediaDetailAdapter adapter;
@@ -78,21 +78,30 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
         return removedItems;
     }
 
-    private MediaDetailPagerFragment() {}; // Constructor calls made to be explicit
-    @SuppressLint("ValidFragment")
-    public MediaDetailPagerFragment(Boolean editable, boolean isFeaturedImage) {
-        this.editable = editable;
-        this.isFeaturedImage = isFeaturedImage;
-        isFromFeaturedRootFragment = false;
+
+    public static MediaDetailPagerFragment newInstance(boolean editable, boolean isFeaturedImage) {
+        MediaDetailPagerFragment mediaDetailPagerFragment = new MediaDetailPagerFragment();
+        Bundle args = new Bundle();
+        args.putBoolean("is_editable", editable);
+        args.putBoolean("is_featured_image", isFeaturedImage);
+        mediaDetailPagerFragment.setArguments(args);
+        return mediaDetailPagerFragment;
     }
 
-    @SuppressLint("ValidFragment")
-    public MediaDetailPagerFragment(Boolean editable, boolean isFeaturedImage, int position) {
-        this.editable = editable;
-        this.isFeaturedImage = isFeaturedImage;
-        isFromFeaturedRootFragment = true;
-        this.position = position;
+    public static MediaDetailPagerFragment newInstance(boolean editable, boolean isFeaturedImage,
+        int position) {
+        MediaDetailPagerFragment mediaDetailPagerFragment = new MediaDetailPagerFragment();
+        Bundle args = new Bundle();
+        args.putBoolean("editable", editable);
+        args.putBoolean("isFeaturedImage", isFeaturedImage);
+        args.putInt("current-page", position);
+        mediaDetailPagerFragment.setArguments(args);
+        return mediaDetailPagerFragment;
     }
+
+    private MediaDetailPagerFragment() {}; // Constructor calls made to be explicit
+
+   
 
     @Override
     public View onCreateView(LayoutInflater inflater,
@@ -147,8 +156,9 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
     public void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         if (savedInstanceState != null) {
-            editable = savedInstanceState.getBoolean("editable");
-            isFeaturedImage = savedInstanceState.getBoolean("isFeaturedImage");
+            editable = savedInstanceState.getBoolean("editable", false);
+            isFeaturedImage = savedInstanceState.getBoolean("isFeaturedImage", false);
+            pager.setCurrentItem(savedInstanceState.getInt("current-page"), false);
         }
         setHasOptionsMenu(true);
         initProvider();

--- a/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
@@ -79,22 +79,18 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
     }
 
 
+    /**
+     * This method will create a new instance of MediaDetailPagerFragment and the arguments will be
+     * saved to a bundle which will be later available in the {@link #onCreate(Bundle)}
+     * @param editable
+     * @param isFeaturedImage
+     * @return
+     */
     public static MediaDetailPagerFragment newInstance(boolean editable, boolean isFeaturedImage) {
         MediaDetailPagerFragment mediaDetailPagerFragment = new MediaDetailPagerFragment();
         Bundle args = new Bundle();
         args.putBoolean("is_editable", editable);
         args.putBoolean("is_featured_image", isFeaturedImage);
-        mediaDetailPagerFragment.setArguments(args);
-        return mediaDetailPagerFragment;
-    }
-
-    public static MediaDetailPagerFragment newInstance(boolean editable, boolean isFeaturedImage,
-        int position) {
-        MediaDetailPagerFragment mediaDetailPagerFragment = new MediaDetailPagerFragment();
-        Bundle args = new Bundle();
-        args.putBoolean("editable", editable);
-        args.putBoolean("isFeaturedImage", isFeaturedImage);
-        args.putInt("current-page", position);
         mediaDetailPagerFragment.setArguments(args);
         return mediaDetailPagerFragment;
     }
@@ -158,7 +154,9 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
         if (savedInstanceState != null) {
             editable = savedInstanceState.getBoolean("editable", false);
             isFeaturedImage = savedInstanceState.getBoolean("isFeaturedImage", false);
-            pager.setCurrentItem(savedInstanceState.getInt("current-page"), false);
+            if(null != pager) {
+                pager.setCurrentItem(savedInstanceState.getInt("current-page", 0), false);
+            }
         }
         setHasOptionsMenu(true);
         initProvider();

--- a/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
+++ b/app/src/main/java/fr/free/nrw/commons/media/MediaDetailPagerFragment.java
@@ -80,6 +80,9 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
 
 
     /**
+     * Use this factory method to create a new instance of this fragment using the provided
+     * parameters.
+     * 
      * This method will create a new instance of MediaDetailPagerFragment and the arguments will be
      * saved to a bundle which will be later available in the {@link #onCreate(Bundle)}
      * @param editable
@@ -95,8 +98,9 @@ public class MediaDetailPagerFragment extends CommonsDaggerSupportFragment imple
         return mediaDetailPagerFragment;
     }
 
-    private MediaDetailPagerFragment() {}; // Constructor calls made to be explicit
-
+    public MediaDetailPagerFragment() {
+        // Required empty public constructor
+    };
    
 
     @Override

--- a/app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailPagerFragmentUnitTests.kt
+++ b/app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailPagerFragmentUnitTests.kt
@@ -67,9 +67,9 @@ class MediaDetailPagerFragmentUnitTests {
 
         val activity = Robolectric.buildActivity(SearchActivity::class.java).create().get()
 
-        fragment = MediaDetailPagerFragment(false, true)
-        fragment = MediaDetailPagerFragment(false, false)
-        fragment = MediaDetailPagerFragment(false, false, 0)
+        fragment = MediaDetailPagerFragment.newInstance(false, true);
+        fragment = MediaDetailPagerFragment.newInstance(false, false);
+        fragment = MediaDetailPagerFragment.newInstance(false, false, 0);
         fragmentManager = activity.supportFragmentManager
         val fragmentTransaction: FragmentTransaction = fragmentManager.beginTransaction()
         fragmentTransaction.add(fragment, null)

--- a/app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailPagerFragmentUnitTests.kt
+++ b/app/src/test/kotlin/fr/free/nrw/commons/media/MediaDetailPagerFragmentUnitTests.kt
@@ -69,7 +69,6 @@ class MediaDetailPagerFragmentUnitTests {
 
         fragment = MediaDetailPagerFragment.newInstance(false, true);
         fragment = MediaDetailPagerFragment.newInstance(false, false);
-        fragment = MediaDetailPagerFragment.newInstance(false, false, 0);
         fragmentManager = activity.supportFragmentManager
         val fragmentTransaction: FragmentTransaction = fragmentManager.beginTransaction()
         fragmentTransaction.add(fragment, null)

